central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,503 @@
+package datastore
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/stackrox/rox/central/metrics"
+	processIndicatorStore "github.com/stackrox/rox/central/processindicator/datastore"
+	"github.com/stackrox/rox/central/processlisteningonport/store/postgres"
+	"github.com/stackrox/rox/central/role/resources"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+type datastoreImpl struct {
+	storage            postgres.Store
+	indicatorDataStore processIndicatorStore.DataStore
+}
+
+var (
+	plopSAC = sac.ForResource(resources.DeploymentExtension)
+	log     = logging.LoggerForModule()
+)
+
+func newDatastoreImpl(
+	storage postgres.Store,
+	indicatorDataStore processIndicatorStore.DataStore,
+) *datastoreImpl {
+	return &datastoreImpl{
+		storage:            storage,
+		indicatorDataStore: indicatorDataStore,
+	}
+}
+
+func (ds *datastoreImpl) AddProcessListeningOnPort(
+	ctx context.Context,
+	portProcesses ...*storage.ProcessListeningOnPortFromSensor,
+) error {
+	defer metrics.SetDatastoreFunctionDuration(
+		time.Now(),
+		"ProcessListeningOnPort",
+		"AddProcessListeningOnPort",
+	)
+
+	if !env.PostgresDatastoreEnabled.BooleanSetting() {
+		// PLOP is a Postgres-only feature, do nothing.
+		log.Warnf("Tried to add PLOP not on Postgres, ignore: %+v", portProcesses)
+		return nil
+	}
+
+	if ok, err := plopSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	portProcesses, completedInBatch := normalizePLOPs(portProcesses)
+
+	// TODO ROX-14376: The next two calls, fetchIndicators and fetchExistingPLOPs, have to
+	// be done in a single join query fetching both ProcessIndicator and needed
+	// bits from PLOP.
+	indicatorsMap, indicatorIds, err := ds.fetchIndicators(ctx, portProcesses...)
+	if err != nil {
+		return err
+	}
+
+	existingPLOPMap, err := ds.fetchExistingPLOPs(ctx, indicatorIds)
+	if err != nil {
+		return err
+	}
+
+	plopObjects := []*storage.ProcessListeningOnPortStorage{}
+	for _, val := range portProcesses {
+		indicatorID := ""
+		var processInfo *storage.ProcessIndicatorUniqueKey
+
+		key := getPlopProcessUniqueKey(val)
+
+		if indicator, ok := indicatorsMap[key]; ok {
+			indicatorID = indicator.GetId()
+			log.Debugf("Got indicator %s: %+v", indicatorID, indicator)
+		} else {
```

Comment:
do we ever clean up and/or reconcile the case where there is no indicatorID?  Theoretically that could result in a fair number of collisions with the map and orphaned data in the database.

Reply:
They're getting cleaned up during pruning. There are ideas about reconciliation, but first we wanted to get understanding how frequent this happening in practice.